enclave/asym_keys.c

```diff
@@ -0,0 +1,553 @@
+        if (key_info_local == NULL)
+            OE_RAISE(OE_OUT_OF_MEMORY);
+    }
+
+    /* Now, get the key buffers. */
+    result = oe_get_seal_key_by_policy(
+        policy,
+        key_buffer_local,
+        &key_buffer_size_local,
+        key_info_local,
+        &key_info_size_local);
+
+    if (result != OE_OK)
+        OE_RAISE(result);
+
+    result = OE_OK;
+    *key_buffer = key_buffer_local;
+    *key_buffer_size = key_buffer_size_local;
+    if (key_info)
+    {
+        *key_info = key_info_local;
+        *key_info_size = key_info_size_local;
+    }
+    key_buffer_local = NULL;
+    key_info_local = NULL;
+
+done:
+    if (key_buffer_local != NULL)
+    {
+        oe_secure_zero_fill(key_buffer_local, key_buffer_size_local);
+        oe_free(key_buffer_local);
+    }
+
+    if (key_info_local != NULL)
+    {
+        oe_secure_zero_fill(key_info_local, key_info_size_local);
+        oe_free(key_info_local);
+    }
+
+    return result;
+}
+
+static oe_result_t _load_seal_key(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    uint8_t** key_buffer,
+    size_t* key_buffer_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    uint8_t* key_buffer_local = NULL;
+    size_t key_buffer_size_local = 0;
+    uint8_t temp;
+
+    if (!key_info || !key_buffer || !key_buffer_size)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    /* Call once to get the size. */
+    result =
+        oe_get_seal_key(key_info, key_info_size, &temp, &key_buffer_size_local);
+
+    if (result == OE_OK)
+        OE_RAISE(OE_UNEXPECTED);
+    if (result != OE_BUFFER_TOO_SMALL)
+        OE_RAISE(result);
+
+    /* Allocate and call again. */
+    key_buffer_local = (uint8_t*)oe_malloc(key_buffer_size_local);
+    if (key_buffer_local == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    result = oe_get_seal_key(
+        key_info, key_info_size, key_buffer_local, &key_buffer_size_local);
+
+    if (result != OE_OK)
+        OE_RAISE(result);
+
+    result = OE_OK;
+    *key_buffer = key_buffer_local;
+    *key_buffer_size = key_buffer_size_local;
+    key_buffer_local = NULL;
+
+done:
+    if (key_buffer_local != NULL)
+    {
+        oe_secure_zero_fill(key_buffer_local, key_buffer_size_local);
+        oe_free(key_buffer_local);
+    }
+
+    return result;
+}
+
+static oe_result_t _create_asymmetric_keypair(
+    const oe_asymmetric_key_params_t* key_params,
+    const uint8_t* master_key,
+    size_t master_key_size,
+    oe_ec_private_key_t* private_key,
+    oe_ec_public_key_t* public_key)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    uint8_t key[OE_SHA256_SIZE];
+    uint8_t key_prev[sizeof(key)];
+    oe_ec_type_t type;
+
+    if (!key_params || !master_key || !private_key || !public_key)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    switch (key_params->type)
+    {
+        case OE_ASYMMETRIC_KEY_EC_SECP256P1:
+            type = OE_EC_TYPE_SECP256R1;
+            break;
```

Comment:
Any idea on why this specific type is the only one supported?

Reply:
It's the most common EC curve and the only one supported by our internal crypto API.